Cirrus: Fix obtaining a CI VM

cevich opened the pull request:
Also removed automatic exection of setup_environment.sh since most
people using this script are podman developers (not automation/CI
folks).  If executing the automation scripts is necessary, manual
attendance to required variables like `$TEST_FLAVOR` is mandatory.

cevich commented:
@edsantiago when you have a moment, if you could give this a try?  I spent quite a lot of time modernizing this (still) hacky script to run entirely using rootless podman.  No more sudo requirement :smiley:  I haven't tested out the "new user" one-time setup experience, so I'm curious how it works (or fails) for you.

edsantiago commented:
After ~20m of pulling a 700M image, it just errors out (lots of repetitions of the same thing). The relevant part seems to be:
```
There was a problem refreshing your current auth tokens: ('invalid_grant: Bad Request
```
I don't see a README or obvious code comments explaining what to do, so I'll try again in the morning. Thanks for tackling this.

edsantiago commented:
Friendly suggestion:
```diff

diff --git a/hack/get_ci_vm.sh b/hack/get_ci_vm.sh
index ab667aacf..2e9a95cf1 100755
--- a/hack/get_ci_vm.sh
+++ b/hack/get_ci_vm.sh
@@ -165,10 +165,24 @@ mkdir -p $TMPDIR/.ssh
 mkdir -p {$HOME,$TMPDIR}/.config/gcloud/ssh
 chmod 700 {$HOME,$TMPDIR}/.config/gcloud/ssh $TMPDIR/.ssh

-# Attempt to determine if named 'libpod' gcloud configuration exists
-showrun $PGCLOUD info > $TMPDIR/gcloud-info
-if egrep -q "Account:.*None" $TMPDIR/gcloud-info
-then
+# Returns true if user has run an 'init' and has a valid token for this gcloud
+function has_valid_token() {
+    # Attempt to determine if named 'libpod' gcloud configuration exists
+    # (FIXME: what does this comment mean anyway?)
+    if $PGCLOUD info | egrep -q 'Account:.*None'; then
+        return 1
+    fi
+
+    # It's possible for 'gcloud info' to list expired credentials,
+    # e.g. 'ERROR:  ... invalid grant: Bad Request'
+    if $PGCLOUD auth print-access-token |& grep ERROR; then
+        return 1
+    fi
+
+    return 0
+}
+
+if ! has_valid_token; then
     echo -e "\n${YEL}WARNING: Can't find gcloud configuration for libpod, running init.${NOR}"
     echo -e "         ${RED}Please choose \"#1: Re-initialize\" and \"login\" if asked.${NOR}"
     showrun $PGCLOUD init --project=$PROJECT --console-only --skip-diagnostics
```

Unfortunately I'm still not going to be able to use or even test this script: the `Transferring tarball` step currently was claiming a 56-minute ETA, so I ^C'ed.

cevich commented:
@edsantiago thanks for taking a peek...

> so I'll try again in the morning

Did that message stop spewing after trying again? Your latest comment says you got to the "upload" stage, so that makes me think you progressed pass the gcloud initialization.

hmmm, so it seems there's some big assumptions on available network bandwidth...Valentine hit this as well, but it was more about the VM being located in the us-central zone, and so hard to work with.  I found a workaround there, and the image-download should be a one-time cost.  The upload part is quite a bit trickier though :confused: 

I've found it immensely useful to have a copy of my current local repo. on the remote VM to play with.  Combined with all the CI-env. var. work the script does for you, makes it really easy to tinker on automation-specific stuff, but outside of Cirrus-CI.  Perhaps if instead of uploading a full tarball, the script could leverage the VM's fat-pipe.  For example, by initially cloning the podman repo. directly, and then somehow only uploading the developer's diff with their local changes.

Any suggestions how to do that?  Could a `git push` from the developers workstation, be used to update the clone on the VM?  I'm worried the script would need to make some big assumptions about the local state, like what branch it's on.

edsantiago commented:
>Did that message stop spewing after trying again?

No. See my diff above; I had to instrument the code so it would detect "bad token" and run the re-init code. Only after that did I get to tar-upload.

>Any suggestions how to [push local git tree, changes-only]

Wow, interesting question. It sounds really tricky, figuring out branch state, local changes (committed and uncommitted), and dealing with `gcloud ssh/scp` on top of all that. No, I have no suggestions, sorry!

cevich commented:
> Wow, interesting question. It sounds really tricky, figuring out branch state, local changes (committed and uncommitted), and dealing with gcloud ssh/scp on top of all that. No, I have no suggestions, sorry!

I can get the VM's external IP address, but without monkeying with copying the developers ssh public keys, or using an (insecure) shared key, setting up normal (non-gcloud) `ssh` is a PITA :frowning_face: 

Hmmm...Yeah, tricky...maybe it's  possible to partially solve this.  I can do a repo clone easy enough on the VM, that would provide a starting place.  Then perhaps it's okay to just rely on the developer to get the remote code in-line with their needs.  Let me try that on for size...

cevich commented:
@edsantiago okay how about this, the only initial setup it does now is a full clone of the upstream repository on the VM.  Oh, and I added your suggested "bad token" fix as well.

edsantiago commented:
Hmmm, how about adding an early `podman pull` of the monster-huge image? 

Reason: I just tried running the script, and it just hung. Likely reason is that my podman testing had rmi'ed the monster image, so there was a pull going on that I didn't see. I have it cached (since last night), so I just ^C'ed and `podman load`ed, and all is working now. But the hang gave no hint of what was going on.

cevich commented:
Good idea...

cevich commented:
Updated:  Found a 7x smaller gcloud image to use :smile: added suggested `podman pull...` and also added ability for some personal VM customization through a "hook" script.  @edsantiago I think you'll find this version ***MUCH*** faster :grinning:

edsantiago commented on hack/get_ci_vm.sh:
```diff
@@ -1,49 +1,81 @@
+    show_usage "Clone of containers/podman must be a subdirectory of \$HOME ($HOME)"
+
+# Disable SELinux labeling to allow read-only mounting of repository files
+PGCLOUD="podman run -it --rm --security-opt label=disable -v $TMPDIR:$TMPDIR -v $HOME:/root:ro -v $HOME/.config/gcloud:/root/.config/gcloud -v $HOME/.config/gcloud/ssh:/root/.ssh -v $LIBPODROOT:$LIBPODROOT:ro $GCLOUD_IMAGE gcloud --configuration=libpod --project=$PROJECT"
```
Bind-mounting `$HOME` makes me queasy. Could you keep the old way (mounting a tmpdir), and change the convention so the hook lives in `.config/gcloud`?

cevich replied:
Yeah, I'm not very comfortable with it either.  Yep, that's a fine place for the hook.

cevich commented:
@baude mind taking a peek at this.  It's working a bit different from before, in the interest of both speed and bandwidth use.  On the other hand, it's a lot more flexible with a new hook-script for every developer to customize the VM setup to their own needs.

edsantiago commented:
New image pulls in only two minutes, much nicer, thank you! The script is now working fine for me. LGTM!

cevich commented:
> The script is now working fine for me.

Fantastic, I'm glad to hear it.  This has ***REALLY*** come in handy for debugging automation scripting problems and helping reproduce those "it only happens in CI" problems.  Thanks for the review and helpful feedback.

rhatdan commented:
/approve
/lgtm
